xgi/utils/tensor.py

```diff
@@ -0,0 +1,191 @@
+## Tensor times same vector in all but one (TTSV1) and all but two (TTSV2)
+from collections import defaultdict
+from itertools import combinations
+from math import factorial
+
+import numpy as np
+from numpy import prod
+from scipy.signal import convolve
+from scipy.sparse import coo_array
+from scipy.special import binom as binomial
+
+__all__ = [
+    "pairwise_incidence",
+    "ttsv1",
+    "ttsv2",
+]
+
+
+def pairwise_incidence(edgedict):
+    """Create pairwise adjacency dictionary from hyperedge list dictionary
+
+    Parameters
+    ----------
+    edgedict : dict
+        edge IDs are keys, edges are values
+
+    Returns
+    -------
+    pairs : dict
+        a dictionary with node pairs as keys and the hyperedges they appear in as values
+    """
+    pairs = defaultdict(set)
+    for e, edge in edgedict.items():
+        for i, j in combinations(sorted(edge), 2):
+            pairs[(i, j)].add(e)
+        for n in edge:
+            pairs[(n, n)].add(e)
+    return pairs
+
+
+def banerjee_coeff(l, r):
+    """Return the Banerjee alpha coefficient
+
+    Parameters
+    ----------
+    l : int
+        size of given hyperedge
+    r : int
+        maximum hyperedge size
+
+    Returns
+    -------
+    float
+        the Banerjee coefficient
+    """
+    return sum(((-1) ** j) * binomial(l, j) * (l - j) ** r for j in range(l + 1))
+
+
+def get_gen_coef_subset_expansion(edge_values, node_value, r):
```

Comment:
missing docs also a few functions after this one

Reply:
I will add these.

Reply:
I added an underscore to emphasize that we're not exposing these functions to users. I moved them to the end of the file and added a note that they are helper functions. TBH, I'm not the best person to write docstrings for these functions, but if Ilya wants to give it a go, I'm fine with that.